src/internal/events/events.go

```diff
@@ -159,3 +162,84 @@ func tenantHash(tenID string) string {
 	sum := md5.Sum([]byte(tenID))
 	return fmt.Sprintf("%x", sum)
 }
+
+// ---------------------------------------------------------------------------
+// metrics aggregation
+// ---------------------------------------------------------------------------
+
+type m string
+
+// metrics collection bucket
+const APICall m = "api_call"
+
+// configurations
+const (
+	reportInterval    = 1 * time.Minute
+	retentionDuration = 2 * time.Minute
+)
+
+// NewMetrics embeds a metrics bus into the provided context.  The bus can be
+// utilized with calls like Inc and Since.
+func NewMetrics(ctx context.Context, w io.Writer) (context.Context, func()) {
+	var (
+		// report interval time-bounds metrics into buckets.  Retention
+		// controls how long each interval sticks around.  Neither one controls
+		// logging rates; that's handled by dumpMetrics().
+		sink = metrics.NewInmemSink(reportInterval, retentionDuration)
+		cfg  = metrics.DefaultConfig("corso")
+		sig  = metrics.NewInmemSignal(sink, metrics.DefaultSignal, w)
+	)
+
+	cfg.EnableHostname = false
+	cfg.EnableRuntimeMetrics = false
+
+	gm, err := metrics.NewGlobal(cfg, sink)
+	if err != nil {
+		logger.CtxErr(ctx, err).Error("metrics bus constructor")
+		return ctx, func() { sig.Stop() }
```

Comment:
if we failed to make a metrics object, should we not stop the signal object immediately? Is it going to have anything to dump if we failed?
```suggestion
		logger.CtxErr(ctx, err).Error("metrics bus constructor")
		sig.Stop()
		
		return ctx, func() {}
```

Reply:
also, after looking at `dumpMetrics` should we handle shutting down the metrics signal from there? Presumably once `dumpMetrics` exits there's no other way for the caller to get a handle to things to actually output more metrics is there?

Reply:
> should we not stop the signal object immediately
It's sixes either way, really.  The only way to leak a resource here would be for the caller to panic and not have defer'd the closer.  No harm in protecting against the leak, though, so I'll take the sig.Stop out of the func.

> also, after looking at dumpMetrics should we handle shutting down the metrics signal from there
We could.  Again, it's sixes overall.   I feel safer handling it directly in the controller